### PR TITLE
Add precipitative ML model to fv3fit

### PR DESCRIPTION
**Pull request by mcgibbon**
Added precipitative ML model which explicitly links surface precipitation to heating and moistening profiles.

Added public API:
- Added "precipitative" model type with PrecipitativeHyperparameters
- Added support for "none" as a regularizer type
- Added OptimizerConfig and RegularizerConfig to fv3fit namespace since they are already public API
- Added new classes DenseNetworkConfig and DenseNetwork

Refactored public API:
- Bulleted list of removed or refactored symbols, such as changes to name, type, behavior, argument, etc. Be cautious about doing these and discuss with team more broadly.
- Added optional `datasets` argument to `open_nudge_to_fine` mapper routine which allows altering the zarrs used to construct the dataset, for example to add diagnostic zarrs
- Refactored ArrayPacker init (and only init) to support taking in any Hashable pack_names instead of str only, which get cast to str during init
- breaking change to Dense training configuration 

Significant internal changes:
- Removed unneeded `.load()` call in StackedBatches
- refactored internal _get_input_vector routine used in certain models to separate the scaling as an explicit step after the call, instead of a hidden side-effect
- Added infrastructure supporting explicit exclusion of a model from the "basic" training tests
- Removed unused `seed` argument from `shuffle` routine in `loaders`, as it was causing training to be non-reproducible
- moved DenseHyperparameters to `fv3fit.keras._models.models`
- refactored dense and training routines to use DenseNetworkConfig and DenseNetwork

- [x] Tests added

Resolves #<github issues> [JIRA-TAG]

(Delete this for the commit message)
You are encouraged to check the PR against the [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--A4lKrs~xg7w5Gsb39N6JLNQoAg-IlsYffZgTwyKEylty7NhY).

### CODE REVIEW

File: external/fv3fit/fv3fit/keras/_models/precipitative.py [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+        """
+        if not self._statistics_are_fit:
+            self.fit_statistics(batches[0])
+        return self._fit_loop(batches, validation_batch)
+
+    def _fit_loop(
+        self, batches: Sequence[xr.Dataset], validation_batch: xr.Dataset
+    ) -> None:
+        if self._train_model is None:
+            self._train_model, self._predict_model = self._build_model()
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/precipitative.py:385]: Yes, the train and predict model are constructed of the same layers. This was handy in BPTT training because it let us construct a one-timestep model but train over many timesteps, and here it's useful in theory because we can have predictor outputs which are not used in training.

In practice we need to refactor the diagnostic scripts to not crash if there are outputs with no validation data, so for that reason I'll remove the `predict_model` (right now they're identical apart from the loss function being set or not).

[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/precipitative.py:385]: Ah I was wrong, I wrote a comment reminding me that the train model causes an exception when you try to serialize it because of the loss functions.

I added a docstring to `build_model` saying these share weights (by sharing layers, in this case).

File: external/fv3fit/fv3fit/keras/_models/precipitative.py [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+        if validation_batch is None:
+            validation_data = None
+        else:
+            validation_data = (
+                tuple(validation_batch[name].values for name in self.input_variables),
+                tuple(validation_batch[name].values for name in self.output_variables),
+            )
+        for i_epoch in range(self._epochs):
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/precipitative.py:393]: Yes, I'll fix this with the two new configuration classes/functions mentioned in previous comments.

File: external/fv3fit/fv3fit/keras/_models/precipitative.py [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+                    f"of epoch {i_epoch}..."
+                )
+                history = self._train_model.fit(
+                    x=tuple(batch[name].values for name in self.input_variables),
+                    y=tuple(batch[name].values for name in self.output_variables),
+                    batch_size=self._train_batch_size,
+                    validation_data=validation_data,
+                )
+                loss_over_batches += history.history["loss"]
+                val_loss_over_batches += history.history.get("val_loss", [np.nan])
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/precipitative.py:414]: Yes. I will either remove them, or use them when writing a shared training-loop routine.

File: external/fv3fit/fv3fit/keras/_models/precipitative.py [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+                self.predictor.dump(
+                    os.path.join(str(self._checkpoint_path), f"epoch_{i_epoch}")
+                )
+                logger.info(
+                    f"Saved model checkpoint after epoch {i_epoch} "
+                    f"to {self._checkpoint_path}"
+                )
+
+    @property
+    def predictor(self) -> PureKerasModel:
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/precipitative.py:425]: Yes. I've moved PureKerasModel to `shared.py`.

File: external/fv3fit/fv3fit/keras/_models/recurrent.py [hunk 5]
```diff
@@ -57,20 +57,16 @@ def limit(args):
     return pack([dQ1, dQ2])
 
 
-def _get_input_vector(
-    packer: ArrayPacker,
-    scaler: Optional[LayerStandardScaler],
-    n_window: Optional[int] = None,
-    series: bool = True,
+def get_input_vector(
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/recurrent.py:60]: Done.

File: external/fv3fit/fv3fit/keras/_models/recurrent.py [hunk 6]
```diff
@@ -88,8 +84,6 @@ def _get_input_vector(
             tf.keras.layers.Input(shape=[n_features]) for n_features in features
         ]
     packed = tf.keras.layers.Concatenate()(input_layers)
-    if scaler is not None:
-        packed = scaler.normalize_layer(packed)
     return input_layers, packed
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/recurrent.py:87]: To be honest, I don't see this being a widely-used utility. If I were to move out the concatenation step, the only thing that remains is an `input_layers = [list comprehension]` operation which could be inserted directly into the code instead of calling `get_input_vector`. I shared it in this case to turn two lines of code into one line.

If you think it would be clearer, I can inline this function in `precipitative.py` and return it to `recurrent.py`. In `recurrent.py` it is a lot more important to turn these two lines into one, because this call happens several times.

File: external/fv3fit/fv3fit/sklearn/_random_forest.py [hunk 7]
```diff
@@ -63,6 +63,7 @@ def __init__(
         output_variables: Iterable[str],
         hyperparameters: RandomForestHyperparameters,
     ):
+        super().__init__(sample_dim_name, input_variables, output_variables)
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/sklearn/_random_forest.py:69]: Yes, I'm not a big fan of inheritance because it often leads to symptoms like this. We could remove the inheritance by making Predictor a purely abstract/protocol class and removing its init behavior, instead relying on each predictor to set its `self.input_variables` and `self.output_variables`.

While the inheritance is here, in theory `self._model_wrapper` can modify the input/output variables, and `predictor.__init__` could be updated to do more than just set `self.<things>`, so there isn't a safe way to remove one of these. I'd like to leave it if that's OK.

I do want to set removing init from Predictor as a future PR task. This also came up when talking about DenseModel, which now takes in both `hyperparameters` and `hyperparameters.input_variables` due to Predictor's requirements on its init signature. I can do it or I'd also be happy to give it to someone else who's interested.

File: external/fv3fit/tests/training/test_train.py [hunk 8]
```diff
@@ -7,36 +8,96 @@
...
+
+def get_dataset(model_type, sample_func):
+    if model_type == "precipitative":
+        input_variables = [
+            "air_temperature",
+            "specific_humidity",
+            "pressure_thickness_of_atmospheric_layer",
+            "physics_precip",
+        ]
+        output_variables = ["dQ1", "dQ2", "total_precipitation_rate"]
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/tests/training/test_train.py:93]: I removed this functionality.

File: external/loaders/loaders/batches/_sequences.py [hunk 9]
```diff
@@ -145,18 +144,16 @@ def __len__(self) -> int:
...
     Args:
         sequence:  Input sequence to have access indices shuffled
-        seed: Seed for random number generator used for shuffling
     Returns:
         A new shuffled sequence
     """
-    random = RandomState(seed)
     seq_len = len(sequence)
-    shuffled = random.choice(seq_len, size=seq_len, replace=False).tolist()
+    shuffled = np.random.choice(seq_len, size=seq_len, replace=False).tolist()
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/loaders/loaders/batches/_sequences.py:156]: It doesn't assume anything, but you're correct - similarly to `np.shuffle` you would have to set the seed at some point prior to calling this for the result to be deterministic.

What you probably don't want to do is set the seed just before each time you call this routine, because while it would give you deterministic shuffles, it would also shuffle identically each time it's called.

**Review by mcgibbon — COMMENTED** [review]
Wow 32 reply comments! @brianhenn you've given me some really neat ideas for reducing code duplication here, let me know if you're on board with the plan.

**Comment by mcgibbon** [comment]
Created two PRs to refactor the [dense code](https://github.com/VulcanClimateModeling/fv3net/pull/1346) and [training loop](https://github.com/VulcanClimateModeling/fv3net/pull/1347) into classes.

File: external/fv3fit/fv3fit/keras/_models/precipitative.py [hunk 10]
```diff
@@ -0,0 +1,471 @@
...
+    for name in output_variables:
+        if loss_type == "mse":
+            factor = tf.constant(
+                1.0 / n_outputs / np.mean(std[name].values ** 2), dtype=tf.float32
+            )
+            loss_list.append(multiply_loss_by_factor(tf.losses.mse, factor))
+        elif loss_type == "mae":
+            factor = tf.constant(
+                1.0 / n_outputs / np.mean(std[name].values), dtype=tf.float32
+            )
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/precipitative.py:120]: Thanks! updated.

File: external/fv3fit/fv3fit/keras/_models/precipitative.py [hunk 11]
```diff
@@ -0,0 +1,471 @@
...
+        save_model_checkpoints: if True, save one model per epoch when
+            dumping, under a 'model_checkpoints' subdirectory
+        workers: number of workers for parallelized loading of batches fed into
+            training, defaults to serial loading (1 worker)
+        max_queue_size: max number of batches to hold in the parallel loading queue.
+            Defaults to 8.
+        keras_batch_size: actual batch_size to apply in gradient descent updates,
+            independent of number of samples in each batch in batches; optional,
+            uses 32 if omitted
+        couple_precip_to_dQ1_dQ2: if True, try to recover behavior of Dense model type
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/precipitative.py:168]: Yes, updated.

File: external/fv3fit/fv3fit/keras/_models/precipitative.py [hunk 12]
```diff
@@ -0,0 +1,471 @@
...
+        workers: int = 1,
+        max_queue_size: int = 8,
+        depth: int = 3,
+        width: int = 16,
+        train_batch_size: Optional[int] = None,
+        optimizer: Optional[tf.keras.optimizers.Optimizer] = None,
+        kernel_regularizer: Optional[tf.keras.regularizers.Regularizer] = None,
+        residual_regularizer: Optional[tf.keras.regularizers.Regularizer] = None,
+        save_model_checkpoints: bool = False,
+        couple_precip_to_dQ1_dQ2: bool = False,
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/precipitative.py:243]: Done.

File: external/fv3fit/fv3fit/keras/_models/models.py [hunk 13]
```diff
@@ -36,10 +49,52 @@
...
+        input_variables: names of variables to use as inputs
+        output_variables: names of variables to use as outputs
+        weights: loss function weights, defined as a dict whose keys are
+            variable names and values are either a scalar referring to the total
+            weight of the variable. Default is a total weight of 1
+            for each variable.
+        normalize_loss: if True (default), normalize outputs by their standard
+            deviation before computing the loss function
+        optimizer_config: selection of algorithm to be used in gradient descent
+        loss: loss function to use, should be 'mse' or 'mae'
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/keras/_models/models.py:68]: I need to add dense_network and training_loop to this docstring.

File: external/fv3fit/fv3fit/keras/_models/models.py [hunk 14]
```diff
@@ -36,10 +49,52 @@
...
+    """
+
+    input_variables: List[str]
+    output_variables: List[str]
+    weights: Optional[Mapping[str, Union[int, float]]] = None
+    normalize_loss: bool = True
+    optimizer_config: OptimizerConfig = dataclasses.field(
+        default_factory=lambda: OptimizerConfig("Adam")
+    )
+    dense_network: DenseNetworkConfig = dataclasses.field(
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/keras/_models/models.py:85]: Since this breaks current configs, can you raise an error with more specific info about this change that says to put the former top level hyperparameters under `dense_network, training_loop` if an old format config is provided?

File: external/fv3fit/fv3fit/keras/_models/shared/training_loop.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+        training, if 1 uses serial loading instead
+    max_queue_size: max number of batches to hold in the parallel loading queue
+    keras_batch_size: actual batch_size to pass to keras model.fit,
+        independent of number of samples in each data batch in batches
+    """
+
+    epochs: int = 3
+    workers: int = 1
+    max_queue_size: int = 8
+    keras_batch_size: int = 16
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/keras/_models/shared/training_loop.py:44]: `DenseModel.fit_kwargs` used to call this `batch_size` while the earlier PrecipitativeModel and this config call it `keras_batch_size`... IMO now that it's in this config it is more clear that it's the "keras" batch size so the prefix is redundant and it'd be less error prone for people converting the old flatter hyperparameters to the newer version if it is kept as `batch_size`.

File: external/fv3fit/fv3fit/keras/_models/shared/dense_network.py [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+            if self.spectral_normalization:
+                hidden_layer = tfa.layers.SpectralNormalization(
+                    hidden_layer, name=f"spectral_norm_{label}_{i}"
+                )
+            x = hidden_layer(x)
+            hidden_outputs.append(x)
+        output = tf.keras.layers.Dense(
+            n_features_out,
+            activation="linear",
+            # activity_regularizer=self.output_activity_regularizer.instance,
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/keras/_models/shared/dense_network.py:75]: Can you just do a check on the PrecipitativeHyperparameters.DenseModelConfig to check that this is not set in addition to the residual regularizer for that particular model? And then uncomment this so that generic DenseModels still can have an activity_regularizer.

**Review by AnnaKwa — APPROVED** [review]
Thanks Jeremy, I'm approving with a few small requests that are pretty straightforward.

File: external/fv3fit/fv3fit/keras/_models/models.py [hunk 17]
```diff
@@ -81,14 +85,13 @@ class DenseHyperparameters(Hyperparameters):
...
-    epochs: int = 3
+    training_loop: TrainingLoopConfig = dataclasses.field(
+        default_factory=TrainingLoopConfig
+    )
     loss: str = "mse"
     save_model_checkpoints: bool = False
     nonnegative_outputs: bool = False
 
-    # TODO: remove fit_kwargs by fixing how validation data is passed
-    fit_kwargs: Optional[dict] = None
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/models.py]: Done.

File: external/fv3fit/fv3fit/keras/_models/models.py [hunk 18]
```diff
@@ -180,34 +214,16 @@ def fit(
...
         """Fits a model using data in the batches sequence
-        
-        If batch_size is provided as a kwarg, the list of values is for each batch fit.
-        e.g. {"loss":
-            [[epoch0_batch0_loss, epoch0_batch1_loss],
-            [epoch1_batch0_loss, epoch1_batch1_loss]]}
-        If not batch_size is not provided, a single loss per epoch is recorded.
-        e.g. {"loss": [[epoch0_loss], [epoch1_loss]]}
+
+        Makes use of configuration in DenseHyperparameters.training_loop
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/models.py:222]: Thanks!

File: external/fv3fit/fv3fit/keras/_models/models.py (hunk 14 above)
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/models.py:85]: This is a really good point, but I don’t think I have time to add that today. I found out that if you use an old config, it actually raises no error, because we’re not using dacite in strict mode. Is this OK to either not add, or add in a PR where we fix dacite to use strict mode in fv3fit?

File: external/fv3fit/fv3fit/keras/_models/shared/training_loop.py (hunk 15 above)
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/shared/training_loop.py:44]: Renamed it to `batch_size`.

File: external/fv3fit/fv3fit/keras/_models/shared/dense_network.py (hunk 16 above)
Comment (mcgibbon) [reply external/fv3fit/fv3fit/keras/_models/shared/dense_network.py:75]: This was supposed to be removed, I've now removed it (and added docstrings for DenseNetworkConfig).

**Review by brianhenn — APPROVED** [review]
Thanks @mcgibbon this will be fun to try out
